build/xtask/src/dist.rs

```diff
@@ -2138,6 +2156,95 @@ fn write_srec(
     Ok(())
 }
 
+fn write_elf(
+    sections: &BTreeMap<AbiSize, LoadSegment>,
+    kentry: AbiSize,
+    out: &Path,
+) -> Result<()> {
+    use goblin::container::{Container, Ctx, Endian};
+    use goblin::elf::header::{Header, EM_RISCV, ET_EXEC};
+    use goblin::elf::program_header::{PF_R, PF_W, PF_X, PT_LOAD};
+    use goblin::elf::ProgramHeader;
+    use scroll::Pwrite;
+
+    let ctx = Ctx::new(Container::Big, Endian::Little);
+
+    // Generate all the program headers and collect all the sections together.
+    let mut program_headers = Vec::new();
+    let mut sections_data = Vec::new();
+    for (&base, sec) in sections {
+        if sec.data.is_empty() {
+            continue;
+        }
+
+        // Incredibly simple algorithm that attempts to align where this program section appears in
+        // the file (what the loop adjusts) with its location in virtual address space (which was
+        // fixed by the compiler already).
+        while ((base as usize) & 0xfff) != (sections_data.len() & 0xfff) {
+            sections_data.push(0);
+        }
+
+        program_headers.push(ProgramHeader {
+            p_type: PT_LOAD,
+            p_flags: PF_X | PF_W | PF_R,
```

Comment:
I had raised this [bug](http://go/ji/SW-474) to track how we want to set the flags differently for different regions.